File: tool/src/main/kotlin/dev/tyler/grimoire/rules/Dice.kt
package dev.tyler.grimoire.rules

/** Thrown for a malformed or out-of-range dice expression (the oracle's `DiceError`). */
class DiceException(message: String) : IllegalArgumentException(message)

/**
 * mulberry32, bit-for-bit with pipeline/reference/dice.py and the JavaScript original (ADR-0006):
 *
 *     a = (a + 0x6D2B79F5) | 0
 *     t = imul(a ^ (a >>> 15), 1 | a)
 *     t = (t + imul(t ^ (t >>> 7), 61 | t)) ^ t
 *     return (t ^ (t >>> 14)) >>> 0
 *
 * Kotlin `Int` addition and multiplication wrap exactly like `| 0` and `imul`; `ushr` is `>>>`. A die is
 * `floor(u32 / 2^32 * sides) + 1`, computed with one `Long` multiply so no floating point is involved.
 * fixtures/rng.json pins the first outputs for six seeds.
 */
class Mulberry32(seed: Int) {
    /** Only the low 32 bits of a seed matter (the oracle masks with 0xFFFFFFFF). */
    constructor(seed: Long) : this(seed.toInt())

    private var state: Int = seed

    /** The next raw 32-bit word; read it as unsigned with `toLong() and 0xFFFFFFFFL`. */
    fun nextU32(): Int {
        state += 0x6D2B79F5
        var t = (state xor (state ushr 15)) * (1 or state)
        t = (t + ((t xor (t ushr 7)) * (61 or t))) xor t
        return t xor (t ushr 14)
    }

    fun die(sides: Int): Int = ((nextU32().toLong() and 0xFFFFFFFFL) * sides ushr 32).toInt() + 1
}

/** One term of a parsed expression; [sign] is +1 or -1. */
sealed interface Term {
    val sign: Int
}

data class DiceTerm(
    val count: Int,
    val sides: Int,
    /** Number of dice kept for `kh`/`kl`, or null to keep them all. */
    val keep: Int? = null,
    val keepHigh: Boolean = true,
    override val sign: Int = 1,
) : Term {
    fun text(): String = buildString {
        append(count).append('d').append(sides)
        if (keep != null) append(if (keepHigh) "kh" else "kl").append(keep)
    }
}

data class ConstTerm(val value: Int, override val sign: Int = 1) : Term

/**
 * The outcome of one expression: [rolls] holds every die per dice term in the order drawn, [kept] the
 * dice that counted (sorted when a `kh`/`kl` applied), [natural] the kept d20 of a single-d20 check for
 * CRIT / MISS display. [expression] is the normalised text (`d6` renders as `1d6`).
 */
data class Roll(
    val expression: String,
    val seed: Int,
    val terms: List<Term>,
    val rolls: List<List<Int>>,
    val kept: List<List<Int>>,
    val total: Int,
    val natural: Int?,
)

/** Advantage and disadvantage are a rewrite of the d20 term, never tokens of the grammar. */
enum class Advantage { NONE, ADV, DIS }

/**
 * Dice notation — a deliberate subset of Roll20 notation, everything a 5E sheet needs and nothing that
 * invites a keyboard:
 *
 *     expr := term (("+" | "-") term)*
 *     term := dice | INT
 *     dice := [INT] "d" INT [("kh" | "kl") INT]
 *
 * Spaces are ignored and letters are case-insensitive. Sides are one of 2, 3, 4, 6, 8, 10, 12, 20, 100;
 * 1 <= count <= 100; 1 <= keep <= count. Dice are drawn left to right, term by term, from one stream.
 */
object Dice {
    val VALID_SIDES: Set<Int> = setOf(2, 3, 4, 6, 8, 10, 12, 20, 100)
    const val MAX_DICE = 100

    /** Splits "-1d4+3" into "-1d4", "+3"; a leftover character (as in "++1" or "1d20+") means malformed. */
    private val chunk = Regex("[+-]?[^+-]+")
    private val termPattern = Regex("^(?:(\\d+)?d(\\d+)(?:(kh|kl)(\\d+))?|(\\d+))$")

    fun parse(expression: String): List<Term> {
        val text = expression.replace(" ", "").lowercase()
        if (text.isEmpty()) throw DiceException("empty expression")
        val tokens = chunk.findAll(text).map { it.value }.toList()
        if (tokens.isEmpty() || tokens.joinToString("") != text) throw DiceException("malformed expression '$expression'")
        val terms = ArrayList<Term>(tokens.size)
        for (token in tokens) {
            val sign = if (token.startsWith("-")) -1 else 1
            val body = token.trimStart('+', '-')
            val match = termPattern.matchEntire(body) ?: throw DiceException("bad term '$body' in '$expression'")
            val constant = match.groups[5]?.value
            if (constant != null) {
                terms += ConstTerm(number(constant, expression), sign)
                continue
            }
            val count = match.groups[1]?.value?.let { number(it, expression) } ?: 1
            val sides = number(match.groups[2]!!.value, expression)
            if (sides !in VALID_SIDES) throw DiceException("unsupported die d$sides")
            if (count !in 1..MAX_DICE) throw DiceException("dice count $count out of range 1..$MAX_DICE")
            var keep: Int? = null
            var keepHigh = true
            val keepKind = match.groups[3]?.value
            if (keepKind != null) {
                keep = number(match.groups[4]!!.value, expression)
                keepHigh = keepKind == "kh"
                if (keep !in 1..count) throw DiceException("keep $keep out of range 1..$count")
            }
            terms += DiceTerm(count, sides, keep, keepHigh, sign)
        }
        return terms
    }

    /** The oracle's integers are unbounded; here a number that does not fit an Int is simply invalid. */
    private fun number(digits: String, expression: String): Int =
        digits.toIntOrNull() ?: throw DiceException("number $digits too large in '$expression'")

    fun render(terms: List<Term>): String = buildString {
        terms.forEachIndexed { index, term ->
            if (index == 0) {
                if (term.sign < 0) append('-')
            } else {
                append(if (term.sign < 0) '-' else '+')
            }
            append(
                when (term) {
                    is DiceTerm -> term.text()
                    is ConstTerm -> term.value.toString()
                },
            )
        }
    }

    /** Rolls [expression] on a fresh generator for [seed], or on a caller-supplied stream. */
    fun roll(expression: String, seed: Int, rng: Mulberry32 = Mulberry32(seed)): Roll {
        val terms = parse(expression)
        val rolls = ArrayList<List<Int>>()
        val kept = ArrayList<List<Int>>()
        var total = 0
        for (term in terms) {
            when (term) {
                is ConstTerm -> total += term.sign * term.value
                is DiceTerm -> {
                    val dice = List(term.count) { rng.die(term.sides) }
                    val counted = when {
                        term.keep == null -> dice
                        term.keepHigh -> dice.sortedDescending().take(term.keep)
                        else -> dice.sorted().take(term.keep)
                    }
                    rolls += dice
                    kept += counted
                    total += term.sign * counted.sum()
                }
            }
        }
        val natural = terms.filterIsInstance<DiceTerm>().zip(kept)
            .firstOrNull { (term, counted) -> term.sides == 20 && counted.size == 1 }
            ?.second?.first()
        return Roll(render(terms), seed, terms, rolls, kept, total, natural)
    }

    /** Rolls several expressions from ONE stream — the Turn screen's attack-then-damage tap. */
    fun rollMany(expressions: List<String>, seed: Int): List<Roll> {
        val rng = Mulberry32(seed)
        return expressions.map { roll(it, seed, rng) }
    }

    /**
     * Rewrites the first plain `1d20` term as `2d20kh1` (advantage) or `2d20kl1` (disadvantage).
     * [Advantage.NONE] returns the expression exactly as given, unparsed.
     */
    fun withAdvantage(expression: String, mode: Advantage): String {
        if (mode == Advantage.NONE) return expression
        var done = false
        val out = parse(expression).map { term ->
            if (!done && term is DiceTerm && term.sides == 20 && term.count == 1 && term.keep == null) {
                done = true
                DiceTerm(2, 20, 1, mode == Advantage.ADV, term.sign)
            } else {
                term
            }
        }
        if (!done) throw DiceException("no 1d20 term to apply advantage to")
        return render(out)
    }

    /**
     * Critical-hit damage: every dice term is rolled twice and the modifiers added once, so the count
     * (and any keep) of every dice term doubles while constants stay: `1d8+3` becomes `2d8+3`.
     */
    fun withCritical(expression: String): String = render(
        parse(expression).map { term ->
            if (term is DiceTerm) {
                if (term.count * 2 > MAX_DICE) throw DiceException("critical rewrite exceeds the dice cap")
                DiceTerm(term.count * 2, term.sides, term.keep?.let { it * 2 }, term.keepHigh, term.sign)
            } else {
                term
            }
        },
    )

    /** Lowest and highest possible totals; a kept term counts only its kept dice. */
    fun bounds(expression: String): IntRange {
        var lo = 0
        var hi = 0
        for (term in parse(expression)) {
            when (term) {
                is ConstTerm -> {
                    lo += term.sign * term.value
                    hi += term.sign * term.value
                }
                is DiceTerm -> {
                    val n = term.keep ?: term.count
                    if (term.sign < 0) {
                        lo -= n * term.sides
                        hi -= n
                    } else {
                        lo += n
                        hi += n * term.sides
                    }
                }
            }
        }
        return lo..hi
    }

    /** Expected total, treating a kept term as that many plain dice (used for "take the average" HP). */
    fun average(expression: String): Double {
        var total = 0.0
        for (term in parse(expression)) {
            when (term) {
                is ConstTerm -> total += term.sign * term.value
                is DiceTerm -> {
                    val n = term.keep ?: term.count
                    total += term.sign * n * (term.sides + 1) / 2.0
                }
            }
        }
        return total
    }
}

File: pipeline/reference/dice.py
"""Dice notation + deterministic RNG.

Grammar (a deliberate subset of Roll20 notation — everything a 5E sheet needs, nothing
that invites a keyboard):

    expr   := term (("+" | "-") term)*
    term   := dice | INT
    dice   := [INT] "d" INT [("kh" | "kl") INT]      e.g. 2d20kh1, d6, 8d6, 1d8+3

- Whitespace is ignored. Case-insensitive.
- Sides ∈ {2,3,4,6,8,10,12,20,100}; 1 ≤ count ≤ 100; keep ≤ count.
- `adv`/`dis` are NOT tokens: the caller rewrites the d20 term to 2d20kh1 / 2d20kl1
  (see `with_advantage`). This keeps the grammar tiny and the display honest.

RNG: mulberry32 (32-bit), the same generator Tyler's Sudoku tool ported bit-for-bit, so
Kotlin can reproduce every fixture with plain `Int` arithmetic:

    a = (a + 0x6D2B79F5) | 0
    t = imul(a ^ (a >>> 15), 1 | a)
    t = (t + imul(t ^ (t >>> 7), 61 | t)) ^ t
    return (t ^ (t >>> 14)) >>> 0            # uint32
    die = floor(uint32 / 2^32 * sides) + 1

Kotlin translation notes: `>>>` is `ushr`; `imul` is plain `Int` multiplication (wraps);
`>>> 0` means "treat as unsigned" → `toUInt().toLong()` or `toLong() and 0xFFFFFFFFL`.
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field

MASK32 = 0xFFFFFFFF
VALID_SIDES = (2, 3, 4, 6, 8, 10, 12, 20, 100)
MAX_DICE = 100


class Mulberry32:
    """Deterministic 32-bit PRNG. `seed` is any int; only the low 32 bits matter."""

    def __init__(self, seed: int):
        self.state = seed & MASK32

    def next_u32(self) -> int:
        self.state = (self.state + 0x6D2B79F5) & MASK32
        a = self.state
        t = ((a ^ (a >> 15)) * (1 | a)) & MASK32
        t = ((t + ((t ^ (t >> 7)) * (61 | t))) ^ t) & MASK32
        return (t ^ (t >> 14)) & MASK32

    def die(self, sides: int) -> int:
        return (self.next_u32() * sides >> 32) + 1  # == floor(u32/2^32 * sides) + 1, exactly, no floats


@dataclass(frozen=True)
class DiceTerm:
    count: int
    sides: int
    keep: int | None = None  # number kept
    keep_high: bool = True
    sign: int = 1

    def text(self) -> str:
        s = f"{self.count}d{self.sides}"
        if self.keep is not None:
            s += ("kh" if self.keep_high else "kl") + str(self.keep)
        return s


@dataclass(frozen=True)
class ConstTerm:
    value: int
    sign: int = 1


@dataclass
class Roll:
    expression: str
    terms: list  # DiceTerm | ConstTerm
    rolls: list[list[int]] = field(default_factory=list)  # per dice term, raw dice in roll order
    kept: list[list[int]] = field(default_factory=list)   # per dice term, kept dice
    total: int = 0

    @property
    def is_d20(self) -> bool:
        return any(isinstance(t, DiceTerm) and t.sides == 20 for t in self.terms)

    @property
    def natural(self) -> int | None:
        """The kept d20 value for a single-d20 check (for crit / fumble display)."""
        for t, k in zip((t for t in self.terms if isinstance(t, DiceTerm)), self.kept):
            if t.sides == 20 and len(k) == 1:
                return k[0]
        return None


class DiceError(ValueError):
    pass


_TERM_RE = re.compile(r"^(?:(\d+)?d(\d+)(?:(kh|kl)(\d+))?|(\d+))$", re.IGNORECASE)


def parse(expression: str) -> list:
    text = expression.replace(" ", "").lower()
    if not text:
        raise DiceError("empty expression")
    # Split into signed chunks: "-1d4+3" → ["-1d4", "+3"]. Rejects "++", trailing "+", etc.
    tokens = re.findall(r"[+-]?[^+-]+", text)
    if "".join(tokens) != text or not tokens:
        raise DiceError(f"malformed expression {expression!r}")
    chunks: list[tuple[int, str]] = []
    for tok in tokens:
        sgn = -1 if tok.startswith("-") else 1
        chunks.append((sgn, tok.lstrip("+-")))
    terms: list = []
    for sgn, chunk in chunks:
        m = _TERM_RE.match(chunk)
        if not m:
            raise DiceError(f"bad term {chunk!r} in {expression!r}")
        if m.group(5) is not None:
            terms.append(ConstTerm(int(m.group(5)), sgn))
            continue
        count = int(m.group(1)) if m.group(1) else 1
        sides = int(m.group(2))
        if sides not in VALID_SIDES:
            raise DiceError(f"unsupported die d{sides}")
        if not 1 <= count <= MAX_DICE:
            raise DiceError(f"dice count {count} out of range 1..{MAX_DICE}")
        keep = None
        keep_high = True
        if m.group(3):
            keep = int(m.group(4))
            keep_high = m.group(3) == "kh"
            if not 1 <= keep <= count:
                raise DiceError(f"keep {keep} out of range 1..{count}")
        terms.append(DiceTerm(count, sides, keep, keep_high, sgn))
    return terms


def with_advantage(expression: str, mode: str) -> str:
    """Rewrite the first plain 1d20 term as 2d20kh1 (adv) or 2d20kl1 (dis)."""
    if mode not in ("adv", "dis"):
        return expression
    terms = parse(expression)
    out = []
    done = False
    for t in terms:
        if isinstance(t, DiceTerm) and t.sides == 20 and t.count == 1 and t.keep is None and not done:
            t = DiceTerm(2, 20, 1, mode == "adv", t.sign)
            done = True
        out.append(t)
    if not done:
        raise DiceError("no 1d20 term to apply advantage to")
    return render(out)


def with_critical(expression: str) -> str:
    """Critical hit damage: roll all of the attack's damage DICE twice, add modifiers once
    (SRD 5.1 "Critical Hits"). Implemented as a rewrite doubling every dice term's count:
    "1d8+3" → "2d8+3", "2d6+1d4+2" → "4d6+2d4+2". Keep terms are doubled in both count and
    keep so `kh`/`kl` semantics survive ("2d20kh1" → "4d20kh2")."""
    out = []
    for t in parse(expression):
        if isinstance(t, DiceTerm):
            if t.count * 2 > MAX_DICE:
                raise DiceError("critical rewrite exceeds the dice cap")
            t = DiceTerm(t.count * 2, t.sides, None if t.keep is None else t.keep * 2, t.keep_high, t.sign)
        out.append(t)
    return render(out)


def roll_many(expressions: list[str], seed: int) -> list["Roll"]:
    """Roll several expressions from ONE Mulberry32(seed) stream, in order — the attack
    (to-hit, then damage) pair on the Turn screen uses this so a single seed reproduces
    the whole tap in tests."""
    rng = Mulberry32(seed)
    return [roll(e, seed, rng) for e in expressions]


def render(terms: list) -> str:
    parts = []
    for i, t in enumerate(terms):
        body = t.text() if isinstance(t, DiceTerm) else str(t.value)
        if i == 0:
            parts.append(("-" if t.sign < 0 else "") + body)
        else:
            parts.append(("-" if t.sign < 0 else "+") + body)
    return "".join(parts)


def roll(expression: str, seed: int, rng: Mulberry32 | None = None) -> Roll:
    """Roll `expression` with a fresh Mulberry32(seed) (or a caller-supplied stream)."""
    terms = parse(expression)
    r = rng or Mulberry32(seed)
    result = Roll(expression=render(terms), terms=terms)
    total = 0
    for t in terms:
        if isinstance(t, ConstTerm):
            total += t.sign * t.value
            continue
        dice = [r.die(t.sides) for _ in range(t.count)]
        if t.keep is None:
            kept = list(dice)
        else:
            ordered = sorted(dice, reverse=t.keep_high)
            kept = ordered[: t.keep]
        result.rolls.append(dice)
        result.kept.append(kept)
        total += t.sign * sum(kept)
    result.total = total
    return result


def bounds(expression: str) -> tuple[int, int]:
    lo = hi = 0
    for t in parse(expression):
        if isinstance(t, ConstTerm):
            lo += t.sign * t.value
            hi += t.sign * t.value
        else:
            n = t.keep if t.keep is not None else t.count
            a, b = n * 1, n * t.sides
            if t.sign < 0:
                a, b = -b, -a
            lo += a
            hi += b
    return lo, hi


def average(expression: str) -> float:
    """Expected value, ignoring keep-high/low (used for 'take the average' HP)."""
    total = 0.0
    for t in parse(expression):
        if isinstance(t, ConstTerm):
            total += t.sign * t.value
        else:
            n = t.keep if t.keep is not None else t.count
            total += t.sign * n * (t.sides + 1) / 2
    return total
